File: Flatten-Binary-Tree-to-Linked-List/Flatten-Binary-Tree-to-Linked-List.cpp

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *head=NULL;
    void flatten(TreeNode *root) {
        if(!root)
            return;
        if(head)
        {
            head->right=root;
            head->left=NULL;
        }
        TreeNode *temp=root->right;
        head=root;
        flatten(root->left);
        flatten(temp);
    }
};

// NonRecursive:
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *head=NULL;
    void flatten(TreeNode *root) {
        while(root)
        {
        // Attatches the right sub-tree to the rightmost leaf of the left sub-tree:
            if(root->left)
            {
                TreeNode *rightMost = root->left;
                while(rightMost->right)
                    rightMost=rightMost->right;
                rightMost->right=root->right;
                // Makes the left sub-tree to the right sub-tree:
                root->right=root->left;
                root->left=NULL;
            }
            // Flatten the rest of the tree:
            root=root->right;
        }
    }
};

 

File: solutions/Flatten-Binary-Tree-to-Linked-List/Flatten-Binary-Tree-to-Linked-List.py


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    lastNode = None
    def flatten(self, root):
        if root==None:
            return
        if self.lastNode!=None:
            self.lastNode.left = None
            self.lastNode.right = root
        self.lastNode = root
        right = root.right
        self.flatten(root.left)
        self.flatten(right)
 